File: src/utils/recipeImageMapper.ts
// Import recipe images
import dosaImg from "@/assets/dosa.jpg";
import idliImg from "@/assets/idli.jpg";
import sambarImg from "@/assets/sambar.jpg";
import rasamImg from "@/assets/rasam.jpg";
import pongalImg from "@/assets/pongal.jpg";
import payasamImg from "@/assets/payasam.jpg";
import vadaImg from "@/assets/vada.jpg";
import chettinadChickenImg from "@/assets/chettinad-chicken.jpg";
import { generateRecipeImage } from "./geminiImageGenerator";

// Import step images
import idliStep1 from "@/assets/steps/idli-step1.jpg";
import idliStep2 from "@/assets/steps/idli-step2.jpg";
import idliStep3 from "@/assets/steps/idli-step3.jpg";
import idliStep4 from "@/assets/steps/idli-step4.jpg";
import idliStep5 from "@/assets/steps/idli-step5.jpg";
import dosaStep1 from "@/assets/steps/dosa-step1.jpg";
import dosaStep2 from "@/assets/steps/dosa-step2.jpg";
import dosaStep3 from "@/assets/steps/dosa-step3.jpg";
import sambarStep1 from "@/assets/steps/sambar-step1.jpg";
import sambarStep2 from "@/assets/steps/sambar-step2.jpg";
import sambarStep3 from "@/assets/steps/sambar-step3.jpg";
import sambarStep4 from "@/assets/steps/sambar-step4.jpg";
import pongalStep1 from "@/assets/steps/pongal-step1.jpg";
import pongalStep2 from "@/assets/steps/pongal-step2.jpg";
import pongalStep3 from "@/assets/steps/pongal-step3.jpg";
import chickenStep1 from "@/assets/steps/chicken-step1.jpg";
import chickenStep2 from "@/assets/steps/chicken-step2.jpg";
import chickenStep3 from "@/assets/steps/chicken-step3.jpg";
import vadaStep1 from "@/assets/steps/vada-step1.jpg";
import vadaStep2 from "@/assets/steps/vada-step2.jpg";
import vadaStep3 from "@/assets/steps/vada-step3.jpg";
import payasamStep1 from "@/assets/steps/payasam-step1.jpg";
import payasamStep2 from "@/assets/steps/payasam-step2.jpg";
import payasamStep3 from "@/assets/steps/payasam-step3.jpg";
import rasamStep1 from "@/assets/steps/rasam-step1.jpg";
import rasamStep2 from "@/assets/steps/rasam-step2.jpg";
import rasamStep3 from "@/assets/steps/rasam-step3.jpg";

/**
 * Maps recipe names to available images in the assets folder
 */
const recipeImageMap: Record<string, string> = {
  // Dosa variations
  'dosa': dosaImg,
  'masala dosa': dosaImg,
  'rava dosa': dosaImg,
  'onion dosa': dosaImg,
  
  // Idli variations
  'idli': idliImg,
  'rava idli': idliImg,
  'kanchipuram idli': idliImg,
  
  // Sambar variations
  'sambar': sambarImg,
  'kongu nadu sambar': sambarImg,
  'chola nadu sambar': sambarImg,
  'arachuvitta sambar': sambarImg,
  'kootu sambar': sambarImg,
  'paruppu sambar': sambarImg,
  'keerai sambar': sambarImg,
  'vendakkai sambar': sambarImg,
  'kathrikai sambar': sambarImg,
  'murungakkai sambar': sambarImg,
  
  // Rasam variations
  'rasam': rasamImg,
  'tomato rasam': rasamImg,
  'pepper rasam': rasamImg,
  'lemon rasam': rasamImg,
  'pineapple rasam': rasamImg,
  'jeera rasam': rasamImg,
  'garlic rasam': rasamImg,
  'mysore rasam': rasamImg,
  'paruppu rasam': rasamImg,
  
  // Pongal variations
  'pongal': pongalImg,
  'ven pongal': pongalImg,
  'sweet pongal': pongalImg,
  
  // Payasam/Kheer variations
  'payasam': payasamImg,
  'semiya payasam': payasamImg,
  'aval payasam': payasamImg,
  'rice payasam': payasamImg,
  'wheat payasam': payasamImg,
  'badam payasam': payasamImg,
  'kheer': payasamImg,
  'pal payasam': payasamImg,
  'pradhaman': payasamImg,
  
  // Vada variations
  'vada': vadaImg,
  'medu vada': vadaImg,
  'masala vada': vadaImg,
  'paruppu vada': vadaImg,
  'bonda': vadaImg,
  'urad dal bonda': vadaImg,
  'potato bonda': vadaImg,
  
  // Chicken variations
  'chicken': chettinadChickenImg,
  'chettinad chicken': chettinadChickenImg,
  'madurai chicken': chettinadChickenImg,
  'kongu nadu chicken': chettinadChickenImg,
  'chola nadu chicken': chettinadChickenImg,
  'tanjore chicken': chettinadChickenImg,
  'pandya nadu chicken': chettinadChickenImg,
  'nadu nadu chicken': chettinadChickenImg,
  'chicken curry': chettinadChickenImg,
  'chicken kuzhambu': chettinadChickenImg,
  'chicken peratal': chettinadChickenImg,
  'chicken varuval': chettinadChickenImg,
  'chicken 65': chettinadChickenImg,
  'chicken biryani': chettinadChickenImg,
  
  // Mutton variations (use chicken image as fallback)
  'mutton': chettinadChickenImg,
  'chettinad mutton': chettinadChickenImg,
  'mutton curry': chettinadChickenImg,
  'mutton kuzhambu': chettinadChickenImg,
  'mutton peratal': chettinadChickenImg,
  'mutton biryani': chettinadChickenImg,
  
  // Fish variations (use chicken image as fallback)
  'fish': chettinadChickenImg,
  'fish curry': chettinadChickenImg,
  'fish kuzhambu': chettinadChickenImg,
  'meen kuzhambu': chettinadChickenImg,
  'fish fry': chettinadChickenImg,
  'fish peratal': chettinadChickenImg,
};

/**
 * Simple hash function to convert string to number
 * This ensures consistent image assignment for the same recipe
 */
function hashString(str: string): number {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash; // Convert to 32-bit integer
  }
  return Math.abs(hash);
}

/**
 * All available recipe images
 * Each recipe will get a different image based on hash of its ID/name
 */
const allRecipeImages = [
  dosaImg,
  idliImg,
  sambarImg,
  rasamImg,
  pongalImg,
  payasamImg,
  vadaImg,
  chettinadChickenImg,
];

/**
 * Get image index based on recipe identifier
 * This ensures each recipe gets a unique, consistent image
 */
function getImageIndex(recipeId: string, recipeName: string): number {
  // Combine ID and name for better distribution
  const combined = `${recipeId}-${recipeName}`;
  const hash = hashString(combined);
  return hash % allRecipeImages.length;
}

/**
 * Get image path for a recipe name (synchronous - uses local assets)
 * Uses intelligent matching first, then hash-based assignment for unique images
 */
export function getRecipeImage(recipeName: string, recipeId?: string): string | null {
  if (!recipeName) return null;
  
  const normalizedName = recipeName.toLowerCase().trim();
  
  // Direct match - try exact match first
  if (recipeImageMap[normalizedName]) {
    return recipeImageMap[normalizedName];
  }
  
  // Remove region prefixes for better matching
  const nameWithoutRegion = normalizedName
    .replace(/^(kongu nadu|chola nadu|chettinad|madurai|tanjore|pandya nadu|nanjil nadu|nadu nadu|thondai nadu|pallava nadu)\s+/i, '')
    .trim();
  
  // Try match without region prefix
  if (nameWithoutRegion !== normalizedName && recipeImageMap[nameWithoutRegion]) {
    return recipeImageMap[nameWithoutRegion];
  }
  
  // Partial match - check if recipe name contains any key
  for (const [key, imagePath] of Object.entries(recipeImageMap)) {
    // Check if the key is contained in the recipe name or vice versa
    if (normalizedName.includes(key) || key.includes(normalizedName)) {
      return imagePath;
    }
    // Also check without region prefix
    if (nameWithoutRegion.includes(key) || key.includes(nameWithoutRegion)) {
      return imagePath;
    }
  }
  
  // Category-based fallback matching
  const categoryKeywords: Record<string, string> = {
    'dosa': dosaImg,
    'idli': idliImg,
    'sambar': sambarImg,
    'rasam': rasamImg,
    'pongal': pongalImg,
    'payasam': payasamImg,
    'kheer': payasamImg,
    'vada': vadaImg,
    'bonda': vadaImg,
    'chicken': chettinadChickenImg,
    'mutton': chettinadChickenImg,
    'beef': chettinadChickenImg,
    'fish': chettinadChickenImg,
    'meen': chettinadChickenImg,
    'prawn': chettinadChickenImg,
    'crab': chettinadChickenImg,
    'egg': chettinadChickenImg,
  };
  
  for (const [keyword, imagePath] of Object.entries(categoryKeywords)) {
    if (normalizedName.includes(keyword) || nameWithoutRegion.includes(keyword)) {
      return imagePath;
    }
  }
  
  // Hash-based assignment for unique images per recipe
  // This ensures each recipe gets a different, consistent image
  // Use recipeId if provided for better uniqueness
  const imageIndex = getImageIndex(recipeId || recipeName, recipeName);
  
  return allRecipeImages[imageIndex];
}

/**
 * Get image path for a recipe using Gemini API (async)
 * Falls back to local assets if Gemini generation fails
 */
export async function getRecipeImageFromGemini(
  recipeName: string,
  recipeId?: string,
  region?: string,
  category?: string
): Promise<string | null> {
  if (!recipeName) return null;
  
  // Try to generate image from Gemini API
  try {
    const geminiImage = await generateRecipeImage(recipeName, region, category);
    if (geminiImage) {
      return geminiImage;
    }
  } catch (error) {
    console.error('Error fetching Gemini image:', error);
  }
  
  // Fallback to local image mapping
  return getRecipeImage(recipeName, recipeId);
}

/**
 * Get step images for a recipe
 */
export function getRecipeStepImages(recipeName: string, stepNumber: number): string | null {
  if (!recipeName) return null;
  
  const normalizedName = recipeName.toLowerCase().trim();
  const stepImageMap: Record<string, string[]> = {
    'dosa': [dosaStep1, dosaStep2, dosaStep3],
    'idli': [idliStep1, idliStep2, idliStep3, idliStep4, idliStep5],
    'sambar': [sambarStep1, sambarStep2, sambarStep3, sambarStep4],
    'rasam': [rasamStep1, rasamStep2, rasamStep3],
    'pongal': [pongalStep1, pongalStep2, pongalStep3],
    'payasam': [payasamStep1, payasamStep2, payasamStep3],
    'vada': [vadaStep1, vadaStep2, vadaStep3],
    'chicken': [chickenStep1, chickenStep2, chickenStep3],
  };
  
  // Find matching recipe
  for (const [key, images] of Object.entries(stepImageMap)) {
    if (normalizedName.includes(key) || key.includes(normalizedName)) {
      const stepIndex = stepNumber - 1;
      if (stepIndex >= 0 && stepIndex < images.length) {
        return images[stepIndex];
      }
    }
  }
  
  return null;
}

File: app/utils/recipe_image_mapper.py
"""
Recipe image mapping utilities
"""
import re

# Recipe image mappings
RECIPE_IMAGE_MAP = {
    # Dosa variations
    'dosa': '/assets/dosa.jpg',
    'masala dosa': '/assets/dosa.jpg',
    'rava dosa': '/assets/dosa.jpg',
    'onion dosa': '/assets/dosa.jpg',
    
    # Idli variations
    'idli': '/assets/idli.jpg',
    'rava idli': '/assets/idli.jpg',
    'kanchipuram idli': '/assets/idli.jpg',
    
    # Sambar variations
    'sambar': '/assets/sambar.jpg',
    'kongu nadu sambar': '/assets/sambar.jpg',
    'chola nadu sambar': '/assets/sambar.jpg',
    'arachuvitta sambar': '/assets/sambar.jpg',
    'kootu sambar': '/assets/sambar.jpg',
    'paruppu sambar': '/assets/sambar.jpg',
    'keerai sambar': '/assets/sambar.jpg',
    'vendakkai sambar': '/assets/sambar.jpg',
    'kathrikai sambar': '/assets/sambar.jpg',
    'murungakkai sambar': '/assets/sambar.jpg',
    
    # Rasam variations
    'rasam': '/assets/rasam.jpg',
    'tomato rasam': '/assets/rasam.jpg',
    'pepper rasam': '/assets/rasam.jpg',
    'lemon rasam': '/assets/rasam.jpg',
    'pineapple rasam': '/assets/rasam.jpg',
    'jeera rasam': '/assets/rasam.jpg',
    'garlic rasam': '/assets/rasam.jpg',
    'mysore rasam': '/assets/rasam.jpg',
    'paruppu rasam': '/assets/rasam.jpg',
    
    # Pongal variations
    'pongal': '/assets/pongal.jpg',
    'ven pongal': '/assets/pongal.jpg',
    'sweet pongal': '/assets/pongal.jpg',
    
    # Payasam/Kheer variations
    'payasam': '/assets/payasam.jpg',
    'semiya payasam': '/assets/payasam.jpg',
    'aval payasam': '/assets/payasam.jpg',
    'rice payasam': '/assets/payasam.jpg',
    'wheat payasam': '/assets/payasam.jpg',
    'badam payasam': '/assets/payasam.jpg',
    'kheer': '/assets/payasam.jpg',
    'pal payasam': '/assets/payasam.jpg',
    'pradhaman': '/assets/payasam.jpg',
    
    # Vada variations
    'vada': '/assets/vada.jpg',
    'medu vada': '/assets/vada.jpg',
    'masala vada': '/assets/vada.jpg',
    'paruppu vada': '/assets/vada.jpg',
    'bonda': '/assets/vada.jpg',
    'urad dal bonda': '/assets/vada.jpg',
    'potato bonda': '/assets/vada.jpg',
    
    # Chicken variations
    'chicken': '/assets/chettinad-chicken.jpg',
    'chettinad chicken': '/assets/chettinad-chicken.jpg',
    'madurai chicken': '/assets/chettinad-chicken.jpg',
    'kongu nadu chicken': '/assets/chettinad-chicken.jpg',
    'chola nadu chicken': '/assets/chettinad-chicken.jpg',
    'tanjore chicken': '/assets/chettinad-chicken.jpg',
    'pandya nadu chicken': '/assets/chettinad-chicken.jpg',
    'nadu nadu chicken': '/assets/chettinad-chicken.jpg',
    'chicken curry': '/assets/chettinad-chicken.jpg',
    'chicken kuzhambu': '/assets/chettinad-chicken.jpg',
    'chicken peratal': '/assets/chettinad-chicken.jpg',
    'chicken varuval': '/assets/chettinad-chicken.jpg',
    'chicken 65': '/assets/chettinad-chicken.jpg',
    'chicken biryani': '/assets/chettinad-chicken.jpg',
    
    # Mutton variations
    'mutton': '/assets/chettinad-chicken.jpg',
    'chettinad mutton': '/assets/chettinad-chicken.jpg',
    'mutton curry': '/assets/chettinad-chicken.jpg',
    'mutton kuzhambu': '/assets/chettinad-chicken.jpg',
    'mutton peratal': '/assets/chettinad-chicken.jpg',
    'mutton biryani': '/assets/chettinad-chicken.jpg',
    
    # Fish variations
    'fish': '/assets/chettinad-chicken.jpg',
    'fish curry': '/assets/chettinad-chicken.jpg',
    'fish kuzhambu': '/assets/chettinad-chicken.jpg',
    'meen kuzhambu': '/assets/chettinad-chicken.jpg',
    'fish fry': '/assets/chettinad-chicken.jpg',
    'fish peratal': '/assets/chettinad-chicken.jpg',
}

ALL_RECIPE_IMAGES = [
    '/assets/dosa.jpg',
    '/assets/idli.jpg',
    '/assets/sambar.jpg',
    '/assets/rasam.jpg',
    '/assets/pongal.jpg',
    '/assets/payasam.jpg',
    '/assets/vada.jpg',
    '/assets/chettinad-chicken.jpg',
]

STEP_IMAGE_MAP = {
    'dosa': ['/assets/steps/dosa-step1.jpg', '/assets/steps/dosa-step2.jpg', '/assets/steps/dosa-step3.jpg'],
    'idli': ['/assets/steps/idli-step1.jpg', '/assets/steps/idli-step2.jpg', '/assets/steps/idli-step3.jpg', 
             '/assets/steps/idli-step4.jpg', '/assets/steps/idli-step5.jpg'],
    'sambar': ['/assets/steps/sambar-step1.jpg', '/assets/steps/sambar-step2.jpg', '/assets/steps/sambar-step3.jpg', 
               '/assets/steps/sambar-step4.jpg'],
    'rasam': ['/assets/steps/rasam-step1.jpg', '/assets/steps/rasam-step2.jpg', '/assets/steps/rasam-step3.jpg'],
    'pongal': ['/assets/steps/pongal-step1.jpg', '/assets/steps/pongal-step2.jpg', '/assets/steps/pongal-step3.jpg'],
    'payasam': ['/assets/steps/payasam-step1.jpg', '/assets/steps/payasam-step2.jpg', '/assets/steps/payasam-step3.jpg'],
    'vada': ['/assets/steps/vada-step1.jpg', '/assets/steps/vada-step2.jpg', '/assets/steps/vada-step3.jpg'],
    'chicken': ['/assets/steps/chicken-step1.jpg', '/assets/steps/chicken-step2.jpg', '/assets/steps/chicken-step3.jpg'],
}

CATEGORY_KEYWORDS = {
    'dosa': '/assets/dosa.jpg',
    'idli': '/assets/idli.jpg',
    'sambar': '/assets/sambar.jpg',
    'rasam': '/assets/rasam.jpg',
    'pongal': '/assets/pongal.jpg',
    'payasam': '/assets/payasam.jpg',
    'kheer': '/assets/payasam.jpg',
    'vada': '/assets/vada.jpg',
    'bonda': '/assets/vada.jpg',
    'chicken': '/assets/chettinad-chicken.jpg',
    'mutton': '/assets/chettinad-chicken.jpg',
    'beef': '/assets/chettinad-chicken.jpg',
    'fish': '/assets/chettinad-chicken.jpg',
    'meen': '/assets/chettinad-chicken.jpg',
    'prawn': '/assets/chettinad-chicken.jpg',
    'crab': '/assets/chettinad-chicken.jpg',
    'egg': '/assets/chettinad-chicken.jpg',
}


def hash_string(s: str) -> int:
    """Simple hash function to convert string to number"""
    hash_val = 0
    for char in s:
        hash_val = ((hash_val << 5) - hash_val) + ord(char)
        hash_val = hash_val & hash_val  # Convert to 32-bit integer
    return abs(hash_val)


def get_image_index(recipe_id: str, recipe_name: str) -> int:
    """Get image index based on recipe identifier"""
    combined = f"{recipe_id}-{recipe_name}"
    hash_val = hash_string(combined)
    return hash_val % len(ALL_RECIPE_IMAGES)


def get_recipe_image(recipe_name: str, recipe_id: str = None) -> str:
    """Get image path for a recipe name"""
    if not recipe_name:
        return ALL_RECIPE_IMAGES[0]
    
    normalized_name = recipe_name.lower().strip()
    
    # Direct match
    if normalized_name in RECIPE_IMAGE_MAP:
        return RECIPE_IMAGE_MAP[normalized_name]
    
    # Remove region prefixes
    name_without_region = re.sub(
        r'^(kongu nadu|chola nadu|chettinad|madurai|tanjore|pandya nadu|nanjil nadu|nadu nadu|thondai nadu|pallava nadu)\s+',
        '', normalized_name, flags=re.IGNORECASE
    ).strip()
    
    # Try match without region prefix
    if name_without_region != normalized_name and name_without_region in RECIPE_IMAGE_MAP:
        return RECIPE_IMAGE_MAP[name_without_region]
    
    # Partial match
    for key, image_path in RECIPE_IMAGE_MAP.items():
        if normalized_name in key or key in normalized_name:
            return image_path
        if name_without_region in key or key in name_without_region:
            return image_path
    
    # Category-based fallback
    for keyword, image_path in CATEGORY_KEYWORDS.items():
        if keyword in normalized_name or keyword in name_without_region:
            return image_path
    
    # Hash-based assignment
    image_index = get_image_index(recipe_id or recipe_name, recipe_name)
    return ALL_RECIPE_IMAGES[image_index]


def get_recipe_step_images(recipe_name: str, step_number: int) -> str:
    """Get step images for a recipe"""
    if not recipe_name:
        return ""
    
    normalized_name = recipe_name.lower().strip()
    
    # Find matching recipe
    for key, images in STEP_IMAGE_MAP.items():
        if key in normalized_name or normalized_name in key:
            step_index = step_number - 1
            if 0 <= step_index < len(images):
                return images[step_index]
    
    return ""

